webots_ros2_core/webots_ros2_core/devices/light_sensor_device.py

```diff
@@ -0,0 +1,110 @@
+from rclpy.time import Time
+from webots_ros2_core.math_utils import interpolate_lookup_table
+from .device import Device
+
+
+# https://ieee-dataport.org/open-access/conversion-guide-solar-irradiance-and-lux-illuminance
+IRRADIANCE_TO_ILLUMINANCE = 120
+
+
+class LightSensorDevice(Device):
+    """
+    ROS2 wrapper for Webots LightSensor node.
+
+    Creates suitable ROS2 interface based on Webots LightSensor node instance:
+    https://cyberbotics.com/doc/reference/lightsensor
+
+    It allows the following functinalities:
+    - Publishes range measurements of type `sensor_msgs/Illuminance`
+
+    Args:
+        node (WebotsNode): The ROS2 node.
+        wb_device (LightSensor): Webots node of type LightSensor.
+
+    Kwargs:
+        params (dict): Dictionary with configuration options in format of::
+
+            dict: {
+                'topic_name': str,      # ROS topic name (default will generated from the sensor name)
+                'timestep': int,        # Publish period in ms (default is equal to robot's timestep)
+                'disable': bool,        # Whether to create ROS interface for this sensor (default false)
+                'always_publish': bool, # Publish even if there are no subscribers (default false)
+            }
+
+    """
+
+    def __init__(self, node, wb_device, params=None):
+        self._node = node
+        self._wb_device = wb_device
+        self._last_update = -1
+        self._publisher = None
+
+        # Determine default params
+        params = params or {}
+        self._topic_name = params.setdefault('topic_name', self._create_topic_name(wb_device))
+        self._timestep = params.setdefault('timestep', int(node.robot.getBasicTimeStep()))
+        self._disable = params.setdefault('disable', False)
+        self._always_publish = params.setdefault('always_publish', False)
+
+        # Create topics
+        if not self._disable:
+            self._publisher = self._node.create_publisher(Illuminance, self._topic_name, 1)
+
+    def __get_variance(self, raw_value):
+        table = self._wb_device.getLookupTable()
+
+        # Find relative standard deviation in lookup table
+        relative_std = None
+        for i in range(0, len(table) - 3, 3):
+            if table[i+1] < raw_value < table[i+3+1] or table[i+1] > raw_value > table[i+3+1]:
+                relative_std = table[i+2]
+                break
+        if relative_std is None:
+            if raw_value < table[1]:
+                relative_std = table[2]
+            else:
+                relative_std = table[-1]
+
+        # Calculate variance from the relative standard deviation
+        std = interpolate_lookup_table(raw_value, self._wb_device.getLookupTable()) * IRRADIANCE_TO_ILLUMINANCE * relative_std
+        return std**2
+
+    def step(self):
+        if self._disable:
+            return
+
+        if self._node.robot.getTime() - self._last_update < self._timestep / 1e6:
+            return
```

Comment:
Since this mechanism is going to be used in all the sensors, we can maybe create a Sensor class (that inherits from Device and is inherited by LightSensorDevice,etc.) and that will encapsulate this.

Reply:
Created `SensorDevice`. I guess even more code be moved to `SensorDevice`, but let's see what is needed for the other sensors.